internal/provider/string.go

```diff
@@ -1,219 +1,382 @@
 
-		"min_special": {
-			Description: "Minimum number of special characters in the result. Default value is `0`.",
-			Type:        schema.TypeInt,
-			Optional:    true,
-			Default:     0,
-			ForceNew:    true,
-		},
+			"min_special": {
+				Description: "Minimum number of special characters in the result. Default value is `0`.",
+				Type:        types.Int64Type,
+				Optional:    true,
+				Computed:    true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					tfsdk.RequiresReplace(),
+					defaultInt(0),
+				},
+			},
 
-		"override_special": {
-			Description: "Supply your own list of special characters to use for string generation.  This " +
-				"overrides the default character list in the special argument.  The `special` argument must " +
-				"still be set to true for any overwritten characters to be used in generation.",
-			Type:     schema.TypeString,
-			Optional: true,
-			ForceNew: true,
-		},
+			"override_special": {
+				Description: "Supply your own list of special characters to use for string generation.  This " +
+					"overrides the default character list in the special argument.  The `special` argument must " +
+					"still be set to true for any overwritten characters to be used in generation.",
+				Type:     types.StringType,
+				Optional: true,
+				Computed: true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					tfsdk.RequiresReplace(),
+				},
+			},
 
-		"result": {
-			Description: "The generated random string.",
-			Type:        schema.TypeString,
-			Computed:    true,
-			Sensitive:   sensitive,
-		},
+			"result": {
+				Description: "The generated random string.",
+				Type:        types.StringType,
+				Computed:    true,
+				Sensitive:   sensitive,
+			},
 
-		"id": {
-			Description: idDesc,
-			Computed:    true,
-			Type:        schema.TypeString,
+			"id": {
+				Description: idDesc,
+				Computed:    true,
+				Type:        types.StringType,
+			},
 		},
 	}
 }
 
-func createStringFunc(sensitive bool) func(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	return func(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-		const numChars = "0123456789"
-		const lowerChars = "abcdefghijklmnopqrstuvwxyz"
-		const upperChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
-		var (
-			specialChars = "!@#$%&*()-_=+[]{}<>:?"
-			diags        diag.Diagnostics
+func validatorMinInt(min int64) tfsdk.AttributeValidator {
+	return minIntValidator{min}
+}
+
+type minIntValidator struct {
+	val int64
+}
+
+func (m minIntValidator) Description(context.Context) string {
+	return "MinInt validator ensures that attribute is at least val"
+}
+
+func (m minIntValidator) MarkdownDescription(context.Context) string {
+	return "MinInt validator ensures that attribute is at least `val`"
+}
+
+func (m minIntValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {
+	t := req.AttributeConfig.(types.Int64)
```

Comment:
I'd strongly encourage avoiding a direct type assertion in validators. This can cause a panic if it receives an unexpected type. My recommendation would be to at least use the two-value form of type assertion to return an error on an unexpected type and return early, e.g.

```go
t, ok := req.AttributeConfig.(types.Int64)

if !ok {
  resp.Diagnostics.AddError()
  return
}
```

Or use something similar to the terraform.io documentation here with `tfsdk.ValueAs()`:

```go
var t types.Int64
diags := tfsdk.ValueAs(ctx, req.AttributeConfig, &t)
resp.Diagnostics.Append(diags...)

if diags.HasError() {
  return
}
```

Reference: https://go.dev/doc/effective_go#interface_conversions
Reference: https://www.terraform.io/plugin/framework/validation#creating-attribute-validators